### PR TITLE
Remove redundant `HandshakeHash::alg` field and reduce `unwrap()`ing.

### PR DESCRIPTION
**Pull request by djc**

### CODE REVIEW

File: rustls/src/hash_hs.rs [hunk 1]
```diff
@@ -129,11 +123,14 @@ impl HandshakeHash {
     /// 'handshake_hash' handshake message.  Start this hash
     /// again, with that message at the front.
     pub fn rollup_for_hrr(&mut self) {
-        let old_hash = self.ctx.take().unwrap().finish();
+        let ctx = self.ctx.as_mut().unwrap();
```
Comment (briansmith) [inline rustls/src/hash_hs.rs:126]: I don't understand how this new version of `rollup_for_hrr` is better than the previous version. I think we should revert the changes to `rollup_for_hrr` and leave the rest.

It seems like this could stay as `take()` to simplify the code below.

File: rustls/src/hash_hs.rs [hunk 2]
```diff
@@ -129,11 +123,14 @@ impl HandshakeHash {
...
+
+        let old_ctx = mem::replace(ctx, digest::Context::new(ctx.algorithm()));
+        let old_hash = old_ctx.finish();
         let old_handshake_hash_msg =
             HandshakeMessagePayload::build_handshake_hash(old_hash.as_ref());
 
-        self.ctx = Some(digest::Context::new(self.alg.unwrap()));
+        self.ctx = Some(digest::Context::new(ctx.algorithm()));
```
Comment (briansmith) [inline rustls/src/hash_hs.rs]: It seems wrong that we call `Context::new` twice now.

**Review by briansmith — CHANGES_REQUESTED** [review]

File: rustls/src/hash_hs.rs (hunk 2 above)
Comment (djc) [reply rustls/src/hash_hs.rs]: Yup, that is an oversight. Will fix it soon.

File: rustls/src/hash_hs.rs (hunk 1 above)
Comment (djc) [reply rustls/src/hash_hs.rs:126]: If we take().finish(), I think we no longer have access to the previous ctx to refer to its algorithm. I also prefer this version because it maintains the invariant of self.ctx.is_some() throughout the call.

Comment (briansmith) [reply rustls/src/hash_hs.rs:126]: To be clear, you could `take()` and then get the algorithm and then `finish()`. However, now I understand better what you're trying to do, and I'm fine with it. We just need to remove the duplicate digest context creation.

File: rustls/src/hash_hs.rs (hunk 2 above)
Comment (djc) [reply rustls/src/hash_hs.rs]: Fixed.
